doc/dev/templates.md

```diff
@@ -0,0 +1,143 @@
+# Istio Mixer: Template Developer’s Guide
+
+Templates are a foundational building block of the Mixer architecture. This document
+describes the template format and how to extend Mixer with custom templates.
+
+- [Background](#background)
+- [Template format](#template-format)
+- [Supported field names](#supported-field-names)
+- [Supported field types](#supported-field-types)
+- [Adding a template to Mixer](#adding-a-template-to-mixer)
+- [Template evolution compatibility](#template-evolution-compatibility)
+
+## Background
+
+Mixer is an attribute processing machine. It ingests attributes coming from the proxy and responds by
+triggering calls into a suite of adapters. These adapters in turn communicate with infrastructure
+backends that offer a variety of capabilities. The operator that configures Istio controls what Mixer
+does with incoming attributes and which adapters are called as a result.
+
+![Attribute Processing Machine](./attribute_machine.svg) 
+
+Mixer categorizes adapters based on the type of data they consume. For example, there are metric adapters, logging
+adapters, quota adapters, access control adapters, and more. The type of data Mixer delivers to individual adapters
+depends on each adapter’s category. All adapters of a given category receive the same type of data. For example,
+metric adapters all receive metrics at runtime.
+
+Mixer determines the format of data that each adapter receive using *templates*. A template is a simple protobuf message
+which lays out the kind of data that a specific category of adapter will be given at runtime. The template determines
+the data adapters receive at runtime, as well as the instances operators must create in order to use an adapter.
+The [Adapter Developer's Guide](./adapters.md) explains how templates are automatically transformed into Go
+artifacts that can be used by adapter developers, and into config definitions that can be used by operators. 
+
+Mixer includes a number of [canonical templates](https://github.com/istio/mixer/tree/master/template) which cover
+most of the anticipated workloads that Mixer is expected to be used with. However, the set of supported templates can
+readily be extended in order to support emerging usage scenarios. Note that it’s preferable to use existing templates
+when possible as it tends to deliver a better end-to-end story for the ecosystem. 
```

Comment:
should we also express that this allows the operators configuration to be portable ?